File: src/Problem021AmicableNumbers.java
void main() {
  int limit = 10000;

  int[] divisorsSum = new int[limit];

  for (int i = 1; i < limit; i++) {
    int j = i * 2;
    while (j < limit) {
      divisorsSum[j] += i;
      j += i;
    }
  }

  int totalAmicableSum = 0;

  for (int a = 1; a < limit; a++) {
    int b = divisorsSum[a];
    if (b > a && b < limit) {
      if (divisorsSum[b] == a) {
        totalAmicableSum += a;
        totalAmicableSum += b;
      }
    }
  }

  IO.println("Total: " + totalAmicableSum);
}

File: src/Problem021AmicableNumbers.kt
fun main() {
  val limit = 10000
  val divisorsSum = IntArray(limit) { 0 }

  for (i in 1 until limit) {
    var j = 2 * i

    while (j < limit) {
      divisorsSum[j] += i
      j += i
    }
  }

  var totalAmicableSum = 0

  for (a in 1 until limit) {
    val b = divisorsSum[a]

    if (b in (a + 1)..<limit) {
      if (divisorsSum[b] == a) {
        totalAmicableSum += a
        totalAmicableSum += b
      }
    }
  }

  println("Total: $totalAmicableSum")
}
